### ISSUE TITLE
ParameterConfiguration.Required() doesn't make parameters required

### ISSUE THREAD

**Issue by kjkrum** [OData/AspNetCoreOData#619]
I'm experimenting with actions. Unbound, collection, and entity actions all behave the same way with regard to `ParameterConfiguration.Required()`. The problem could be my expectation of what `Required()` is supposed to do.

```C#
	var action = builder.Action(nameof(UnboundController.UnboundAction));
	action.Parameter<string>("foo").Required();
	action.Parameter<string>("bar").Required();
```
If you underpost to this action, `ModelState.IsValid` is true and the `ODataActionParameters` parameter is non-null. The action will even accept a completely empty JSON object. But if you overpost, `ModelState.IsValid` is false and the `ODataActionParameters` is null.

My expectation is that this action should not accept underposting, but it should accept `{ "foo" : null, "bar" : null }`.

If you make the properties non-nullable instead of required, then underposting causes `ModelState.IsValid` to be false:

```C#
	var action = builder.Action(nameof(UnboundController.UnboundAction));
	action.Parameter<string>("foo").Nullable = false;
	action.Parameter<string>("bar").Nullable = false;
```

Basically, `Required()` seems to do nothing.

**Comment by xuzhg** [comment]
@kjkrum @julealgon See the codes at: https://github.com/OData/ModelBuilder/blob/main/src/Microsoft.OData.ModelBuilder/Operations/ParameterConfiguration.cs#L71-L75

Please allow me to clarify the `nullable` vs `optional` for parameter:

1) nullable means we can or can't accept 'null' value, in the metadata, you will see `nullable=false` attribute, or omitted.
2) optional means the parameter can be omitted or not. in the metadata, you will see an <annotation .... /> element.

**Comment by kjkrum** [comment]
Calling `Required()` does not change the metadata. Calling `Optional` *does* change the metadata by adding this annotation:

```XML
          <Parameter Name="foo" Type="Edm.String">
            <Annotation Term="Org.OData.Core.V1.OptionalParameter" />
          </Parameter>
```
So it seems that parameters are required by default.

My controller action signature is `public IActionResult UnboundAction(ODataActionParameters op)`. I used `ODataActionParameters` because I was following [this doc](https://docs.microsoft.com/en-us/odata/webapi/built-in-routing-conventions) that says that's how controller actions receive params. Route debug shows this as an OData route. If I overpost, causing `ModelState.IsValid` to be false, then serializing the model state shows an error that `op` is required; no mention of `foo` and `bar`. The construction of `ODataActionParameters` does not seem to respect parameters being required.

I changed the action signature to `public IActionResult UnboundAction(string foo, string bar)`. This causes the parameters to be passed as URL parameters instead of as JSON in the post body. This makes them required; the model state is invalid if I don't provide them. Changing the signature also causes the Swagger UI to recognize the params as separate named inputs instead of presenting a generic JSON post body. Swagger still doesn't treat them as required, but that can be remedied with a redundant `[Required]` on the method param.

**Comment by kjkrum** [comment]
My general expectation for what required means was set by [Model Validation in ASP.NET Web API](https://docs.microsoft.com/en-us/aspnet/web-api/overview/formats-and-model-binding/model-validation-in-aspnet-web-api). Overposting does not cause the model state to be invalid; unexpected properties are ignored. Underposting causes the model state to be invalid if and only if the missing property is `[Required]`. Of course, it isn't necessarily wrong for OData to behave differently. It's just surprising.

**Comment by kjkrum** [comment]
I got the behavior I wanted with this controller action signature: `public IActionResult UnboundAction([Required, FromForm] string foo, [Required, FromForm] string bar)`. If I remove `Required` so the Swagger UI allows me to omit a param that's required by the EDM, then the model state is invalid as expected. `FromForm` causes the parameters to be passed in the body as multipart/form-data instead of JSON, but the values are passed to the controller action as expected. Route debug still shows it as an OData route.

**Comment by xuzhg** [comment]
if you don't want it to be OData route, you can change 'UnboundAction' action name to others random name. 

`public IActionResult UnboundAction([Required, FromForm] string foo, [Required, FromForm] string bar)` is not fully supported in 8.x in OData.

**Comment by kjkrum** [comment]
@xuzhg The names of unbound actions don't seem to matter. It's only recognized as an OData route if I decorate it with `[HttpPost($"/{EntityDataModel.RoutePrefix}/{nameof(UnboundAction)}")]`.

In what way is `public IActionResult UnboundAction([Required, FromForm] string foo, [Required, FromForm] string bar)` not supported? It seems to work as expected. Controller actions that receive `ODataActionParameters` do not seem to be fully supported, in that they do not respect parameters being required.

**Comment by julealgon** [comment]
@kjkrum note that, as @xuzhg mentioned, what you are doing is not supported by OData: you can't use `FromForm` and expect OData to detect your endpoint.

Just keep that in mind (that you are in standard MVC territory).

**Comment by kjkrum** [comment]
Conventional routing won't detect it. But if you use explicit attribute routing (which AFAIK you must always do for unbound actions anyway) the route is listed as an OData route in the $odata debug.

**Comment by julealgon** [comment]
> Conventional routing won't detect it. But if you use explicit attribute routing (which AFAIK you must always do for unbound actions anyway) the route is listed as an OData route in the $odata debug.

Oh I'm very surprised to hear that.... I wonder then if you are just hitting some undefined behavior boundary, or if it really is supported at the end of the day.

@xuzhg can you clarify a bit more on this one?

**Comment by kjkrum** [comment]
To be clear, I'd prefer to use `ODataActionParameters` if there's a way to make it treat required parameters as required. I'm looking at [`ODataBodyModelBinder`](https://github.com/OData/AspNetCoreOData/blob/69eec03c7003fe12d92cdc619efdc16781683694/src/Microsoft.AspNetCore.OData/Formatter/ODataBodyModelBinder.cs)...

**Comment by kjkrum** [comment]
I think the issue is in `ODataActionPayloadDeserializer`. It iterates over the properties that are found in the payload and looks up the corresponding `IEdmOperationParameter`. This is backward. It should iterate over `IEdmOperation.Parameters` and find each one in the payload. If a required parameter is not found, it should explode. There's no "is optional" in `IEdmOperationParameter`; optional parameters will be instances of `IEdmOptionalParameter`. That interface includes what the default value should be if the parameter isn't present in the payload.

**Comment by kjkrum** [comment]
This is a lock the doors, order pizza, nobody goes home until it's fixed level bug, and it's been over a year. What is going on at Microsoft?

**Comment by wgibbons1** [comment]
I'm getting similar issues.
When I send a payload with an extra field to my POST endpoint, I get incorrect response error details payload ("Product is required" instead of "Name is required").
When I send a number value instead of string for one of the fields in my payload, the same error message as above (no mention of the actual field).
Very amateurish.

**Comment by julealgon** [comment]
> This is a lock the doors, order pizza, nobody goes home until it's fixed level bug, and it's been over a year. What is going on at Microsoft?

I don't personally understand why you consider this issue _that_ critical... you can just add your own validation temporarily until it is resolved, can't you?

Again, not disagreeing with you on the fact that this _should_ be looked at, but to me this is just not that critical compared to a multitude of other issues in the OData repo.
